crates/paralegal-policy/src/context.rs

```diff
@@ -40,10 +40,118 @@ pub enum EdgeType {
     DataAndControl,
 }
 
+/// Data type representing nodes in the SPDG for a particular controller.
+#[derive(Clone, Copy, PartialEq, Eq, Hash)]
+pub struct Node<'a> {
+    /// ControllerId of the node.
+    pub ctrl_id: ControllerId,
+    /// The data of the node.
+    pub node: NodeType<'a>,
+}
+
+/// Enum identifying the different types of nodes in the SPDG
+#[derive(Clone, Copy, PartialEq, Eq, Hash)]
+pub enum NodeType<'a> {
+    /// Corresponds to [`DataSource::Argument`].
+    CtrlArgument(&'a DataSource),
+
+    /// Corresponds to a [`CallSite`] or [`DataSource::FunctionCall`].
+    CallSite(&'a CallSite),
+
+    /// Corresponds to a [`DataSink::Argument`].
+    CallArgument(&'a DataSink),
+
+    /// Corresponds to a [`DataSink::Return`].
+    Return(&'a DataSink),
+}
+
+impl<'a> From<&'a CallSite> for NodeType<'a> {
+    fn from(cs: &'a CallSite) -> Self {
+        NodeType::CallSite(cs)
+    }
+}
+
+impl<'a> From<&'a DataSink> for NodeType<'a> {
+    fn from(ds: &'a DataSink) -> Self {
+        match ds {
+            DataSink::Argument { .. } => NodeType::CallArgument(ds),
+            DataSink::Return => NodeType::Return(ds),
+        }
+    }
+}
+
+impl<'a> From<&'a CallSiteOrDataSink> for NodeType<'a> {
+    fn from(x: &'a CallSiteOrDataSink) -> Self {
+        match x {
+            CallSiteOrDataSink::CallSite(cs) => cs.into(),
+            CallSiteOrDataSink::DataSink(ds) => ds.into(),
+        }
+    }
+}
+
+impl<'a> From<&'a DataSource> for NodeType<'a> {
+    fn from(ds: &'a DataSource) -> Self {
+        match ds {
+            DataSource::FunctionCall(cs) => NodeType::CallSite(cs),
+            DataSource::Argument(_) => NodeType::CtrlArgument(ds),
+        }
+    }
+}
+
+impl<'a> NodeType<'a> {
+    /// Transform a NodeType into it's corresponding CallSite - only works for CallSites and CallArguments.
+    pub fn as_call_site(&self) -> Option<&'a CallSite> {
```

Comment:
I think I commented this elsewhere but I think these can also just take a plain `self` instead of `&self`

Reply:
done